includes/class-wc-payments-webhook-processing-service.php

```diff
@@ -411,6 +413,67 @@ private function process_webhook_payment_intent_succeeded( $event_body ) {
 			];
 			$this->receipt_service->send_customer_ipp_receipt_email( $order, $merchant_settings, $charges_data[0] );
 		}
+
+		$was_paid_on_woopay = filter_var( $metadata['paid_on_woopay'] ?? false, FILTER_VALIDATE_BOOL );
+		if ( $was_paid_on_woopay ) {
+			$order->add_meta_data( 'is_woopay', true, true );
+
+			// attach_intent_info_to_order.
+			$order->set_transaction_id( $intent_id );
+			$order->update_meta_data( '_intent_id', $intent_id );
+			$order->update_meta_data( '_charge_id', $charge_id );
+			$order->update_meta_data( '_intention_status', $intent_status );
+			$order->update_meta_data( '_payment_method_id', $payment_method );
+			$order->update_meta_data( '_stripe_customer_id', $customer_id );
+			WC_Payments_Utils::set_order_intent_currency( $order, $currency );
+			$order->save();
+
+			// attach_exchange_info_to_order.
+			if ( empty( $charge_id ) ) {
+				return;
+			}
+
+			$currency_store   = strtolower( get_option( 'woocommerce_currency' ) );
+			$currency_order   = strtolower( $order->get_currency() );
+			$currency_account = strtolower( $this->account->get_account_default_currency() );
+
+			// If the default currency for the store is different from the currency for the merchant's Stripe account,
+			// the conversion rate provided by Stripe won't make sense, so we should not attach it to the order meta data
+			// and instead we'll rely on the _wcpay_multi_currency_order_exchange_rate meta key for analytics.
+			if ( $currency_store !== $currency_account ) {
+				return;
+			}
+
+			if ( $currency_order !== $currency_account ) {
+				// We check that the currency used in the order is different than the one set in the WC Payments account
+				// to avoid requesting the charge if not needed.
+				$charge        = $this->payments_api_client->get_charge( $charge_id );
+				$exchange_rate = $charge['balance_transaction']['exchange_rate'] ?? null;
+				if ( isset( $exchange_rate ) ) {
+					$exchange_rate = WC_Payments_Utils::interpret_string_exchange_rate( $exchange_rate, $currency_order, $currency_account );
+					$order->update_meta_data( '_wcpay_multi_currency_stripe_exchange_rate', $exchange_rate );
+					$order->save_meta_data();
+				}
+			}
+
+			// update_order_status_from_intent.
+			switch ( $intent_status ) {
+				case 'succeeded':
+					$this->order_service->mark_payment_completed( $order, $intent_id, $intent_status, $charge_id );
+					break;
+				case 'processing':
+				case 'requires_capture':
+					$this->order_service->mark_payment_authorized( $order, $intent_id, $intent_status, $charge_id );
+					break;
+				case 'requires_action':
+				case 'requires_payment_method':
+					$this->order_service->mark_payment_started( $order, $intent_id, $intent_status, $charge_id );
+					break;
+				default:
+					Logger::error( 'Uncaught payment intent status of ' . $intent_status . ' passed for order id: ' . $order->get_id() );
+					break;
+			}
```

Comment:
These are all direct copies of the `attach_intent_info_to_order`, `attach_exchange_info_to_order`, and `update_order_status_from_intent` functions from the `process_payment_for_order` function in the gateway. If we decide to go with this approach those functions should be moved to a shared class somewhere.

Reply:
I've since refactored these and added each copied function to the order service.